kv/dist_sender.go

```diff
@@ -406,90 +379,141 @@ func (ds *DistSender) sendRPC(trace *tracer.Trace, rangeID proto.RangeID, replic
+// range reaches outside the first descriptor.
+// In case either of the descriptors is discovered stale, the returned closure
+// should be called; it evicts the cache appropriately.
+// Note that `from` and `to` are not necessarily Key and EndKey from a
+// RequestHeader; it's assumed that they've been translated to key addresses
+// already (via KeyAddress).
+func (ds *DistSender) getDescriptors(from, to proto.Key, options lookupOptions) (*proto.RangeDescriptor, *proto.RangeDescriptor, func(), error) {
 	var desc *proto.RangeDescriptor
 	var err error
-	_, isReverseScan := call.Args.(*proto.ReverseScanRequest)
-	if !isReverseScan {
-		desc, err = ds.rangeCache.LookupRangeDescriptor(call.Args.Header().Key, options)
+	var descKey proto.Key
+	if !options.useReverseScan {
+		descKey = from
 	} else {
-		options.useReverseScan = true
-		desc, err = ds.rangeCache.LookupRangeDescriptor(call.Args.Header().EndKey, options)
+		descKey = to
 	}
+	desc, err = ds.rangeCache.LookupRangeDescriptor(descKey, options)
 
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	// Checks whether need to get next range descriptor. If so, returns true
 	// and the key to look up, depending on whether we're in reverse mode.
-	needAnother := func(desc *proto.RangeDescriptor, isReverseScan bool) (proto.Key, bool) {
-		if isReverseScan {
-			return desc.StartKey, call.Args.Header().Key.Less(desc.StartKey)
+	needAnother := func(desc *proto.RangeDescriptor, isReverse bool) (proto.Key, bool) {
+		if isReverse {
+			return desc.StartKey, from.Less(desc.StartKey)
 		}
-		return desc.EndKey, desc.EndKey.Less(call.Args.Header().EndKey)
+		return desc.EndKey, desc.EndKey.Less(to)
 	}
 
 	var descNext *proto.RangeDescriptor
 	// If the request accesses keys beyond the end of this range,
 	// get the descriptor of the adjacent range to address next.
-	if nextKey, ok := needAnother(desc, isReverseScan); ok {
-		if _, ok := call.Reply.(proto.Combinable); !ok {
-			return nil, nil, util.Error("illegal cross-range operation")
-		}
-		// If there's no transaction and op spans ranges, possibly
-		// re-run as part of a transaction for consistency. The
-		// case where we don't need to re-run is if the read
-		// consistency is not required.
-		if call.Args.Header().Txn == nil &&
-			call.Args.Header().ReadConsistency != proto.INCONSISTENT {
-			return nil, nil, &proto.OpRequiresTxnError{}
-		}
+	if nextKey, ok := needAnother(desc, options.useReverseScan); ok {
 		// This next lookup is likely for free since we've read the
 		// previous descriptor and range lookups use cache
 		// prefetching.
 		descNext, err = ds.rangeCache.LookupRangeDescriptor(nextKey, options)
 		if err != nil {
-			return nil, nil, err
+			return nil, nil, nil, err
 		}
 	}
-	return desc, descNext, nil
+	evict := func() {
+		ds.rangeCache.EvictCachedRangeDescriptor(descKey, desc, options.useReverseScan)
+	}
+	return desc, descNext, evict, nil
```

Comment:
I don't think anything is using 'descNext', I don't understand why it is being used here instead of a simple boolean.